vm/src/py_serde.rs

```diff
@@ -0,0 +1,222 @@
+        } else if objtype::isinstance(self.pyobject, &self.vm.ctx.list_type()) {
+            let elements = objsequence::get_elements_list(self.pyobject);
+            serialize_seq_elements(serializer, &elements)
+        } else if objtype::isinstance(self.pyobject, &self.vm.ctx.tuple_type()) {
+            let elements = objsequence::get_elements_tuple(self.pyobject);
+            serialize_seq_elements(serializer, &elements)
+        } else if objtype::isinstance(self.pyobject, &self.vm.ctx.dict_type()) {
+            let dict: PyDictRef = self.pyobject.clone().downcast().unwrap();
+            let pairs: Vec<_> = dict.into_iter().collect();
+            let mut map = serializer.serialize_map(Some(pairs.len()))?;
+            for (key, e) in pairs.iter() {
+                map.serialize_entry(&self.clone_with_object(key), &self.clone_with_object(&e))?;
+            }
+            map.end()
+        } else if self.pyobject.is(&self.vm.get_none()) {
+            serializer.serialize_none()
+        } else {
+            Err(serde::ser::Error::custom(format!(
+                "Object of type '{:?}' is not serializable",
+                self.pyobject.class()
+            )))
+        }
+    }
+}
+
+// This object is used as the seed for deserialization so we have access to the PyContext for type
+// creation
+#[derive(Clone)]
+pub struct PyObjectDeserializer<'c> {
+    vm: &'c VirtualMachine,
+}
+
+impl<'c> PyObjectDeserializer<'c> {
+    pub fn new(vm: &'c VirtualMachine) -> Self {
+        PyObjectDeserializer { vm }
+    }
+}
+
+impl<'de> DeserializeSeed<'de> for PyObjectDeserializer<'de> {
+    type Value = PyObjectRef;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_any(self.clone())
+    }
+}
+
+impl<'de> Visitor<'de> for PyObjectDeserializer<'de> {
+    type Value = PyObjectRef;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a type that can deserialise in Python")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(self.vm.ctx.new_str(value.to_string()))
+    }
+
+    fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(self.vm.ctx.new_str(value))
+    }
+
+    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // The JSON deserializer always uses the i64/u64 deserializers, so we only need to
+        // implement those for now
+        Ok(self.vm.ctx.new_int(value))
+    }
+
+    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // The JSON deserializer always uses the i64/u64 deserializers, so we only need to
+        // implement those for now
```

Comment:
The default implementations of `visit_u32`, `visit_u16`, etc. forward to `visit_u64` anyway, so it’s probably good as is even for other deserializers that might use those.